pkg/cmd/dedicated/deregister/deregister.go

```diff
@@ -0,0 +1,304 @@
+
+	return cmd
+}
+
+func runDeRegisterClusterCmd(opts *options) (err error) {
+	// Set the base URL for the cluster management API
+	err = setListClusters(opts)
+	if err != nil {
+		return err
+	}
+	if len(opts.clusterList) == 0 {
+		return opts.f.Localizer.MustLocalizeError("dedicated.registerCluster.run.noClusterFound")
+	}
+	// TO-DO if client has supplied a cluster id, validate it and set it as the selected cluster without listing getting all clusters
+	if opts.selectedClusterId == "" {
+		err = runClusterSelectionInteractivePrompt(opts)
+		if err != nil {
+			return err
+		}
+	} else {
+		for i := range opts.clusterList {
+			cluster := opts.clusterList[i]
+			if cluster.ID() == opts.selectedClusterId {
+				opts.selectedCluster = cluster
+			}
+		}
+	}
+
+	kafkas, err := getkafkasInCluster(opts)
+	if err != nil {
+		return err
+	}
+
+	err = deleteKafkasPrompt(opts, &kafkas)
+	if err != nil {
+		return err
+	}
+
+	err = deregisterClusterFromKasFleetManager(opts)
+	if err != nil {
+		return err
+	}
+
+	err = removeAddonsFromCluster(opts)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getClusterList(opts *options) (*clustersmgmtv1.ClusterList, error) {
+	conn, err := opts.f.Connection()
+	if err != nil {
+		return nil, err
+	}
+	client, cc, err := conn.API().OCMClustermgmt(opts.clusterManagementApiUrl, opts.accessToken)
+	if err != nil {
+		return nil, err
+	}
+	defer cc()
+	// TO-DO deal with pagination, validate clusters -- must be multi AZ and ready.
+	resource := client.Clusters().List()
+	response, err := resource.Send()
+	if err != nil {
+		return nil, err
+	}
+	clusters := response.Items()
+	return clusters, nil
+}
+
+func setListClusters(opts *options) error {
+	clusters, err := getClusterList(opts)
+	if err != nil {
+		return err
+	}
+	var cls = []clustersmgmtv1.Cluster{}
+	cls = validateClusters(clusters, cls)
+	opts.clusterList = cls
+	return nil
+}
+
+func validateClusters(clusters *clustersmgmtv1.ClusterList, cls []clustersmgmtv1.Cluster) []clustersmgmtv1.Cluster {
+	for _, cluster := range clusters.Slice() {
+		if cluster.State() == clusterReadyState && cluster.MultiAZ() == true {
+			cls = append(cls, *cluster)
+		}
+	}
+	return cls
+}
+
+func getkafkasInCluster(opts *options) ([]kafkamgmtclient.KafkaRequest, error) {
+
+	kafkas := make([]kafkamgmtclient.KafkaRequest, 0)
+
+	conn, err := opts.f.Connection()
+	if err != nil {
+		return kafkas, err
+	}
+
+	api := conn.API()
+
+	// pagination is cringe
+	a := api.KafkaMgmt().GetKafkas(opts.f.Context)
+	a = a.Page(strconv.Itoa(1))
+	a = a.Size(strconv.Itoa(99))
+
+	// deal with response errors at some point
+	kafkaList, _, err := a.Execute()
+	if err != nil {
+		return kafkas, err
+	}
+
+	for _, kafka := range kafkaList.Items {
+		if *kafka.BillingModel == billingModelEnterprise && kafka.ClusterId.IsSet() && *kafka.ClusterId.Get() == opts.selectedCluster.ID() {
```

Comment:
filtering by cluster_id as suggested in https://github.com/redhat-developer/app-services-cli/pull/1812/files#r1112108928 should be enough